Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/CachedQueries/FarePolicyCac.hs

```diff
@@ -0,0 +1,88 @@
+{-
+ Copyright 2022-23, Juspay India Pvt Ltd
+
+ This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License
+
+ as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program
+
+ is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+
+ or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+{-# OPTIONS_GHC -Wno-deprecations #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Storage.CachedQueries.FarePolicyCac where
+
+import qualified "dashboard-helper-api" Dashboard.ProviderPlatform.Merchant as DPM
+import Data.List.NonEmpty
+import Domain.Types.FarePolicy as Domain
+import Kernel.Beam.Functions (FromCacType (..))
+import Kernel.Prelude
+import Kernel.Types.App
+import Kernel.Types.Id
+import Kernel.Utils.Common
+import qualified Storage.Beam.FarePolicy as BeamFP
+import qualified Storage.CachedQueries.FarePolicy.DriverExtraFeeBounds as CQueriesDEFB
+import qualified Storage.CachedQueries.FarePolicy.FarePolicyProgressiveDetails as CQueriesFPPD
+import qualified Storage.CachedQueries.FarePolicy.FarePolicyRentalDetails as CQueriesFPRD
+import qualified Storage.CachedQueries.FarePolicy.FarePolicySlabsDetails.FarePolicySlabsDetailsSlab as CQueriesFPSDS
+import Utils.Common.Cac.ContextConstants
+import Utils.Common.Cac.PrefixConstants as PC
+
+getFarePolicyFromCac :: (CacheFlow m r, EsqDBFlow m r) => [(CacContext, Value)] -> String -> Id FarePolicy -> Int -> m (Maybe FarePolicy)
+getFarePolicyFromCac context tenant id toss = do
+  res :: (Maybe BeamFP.FarePolicy) <- getConfigFromCac context tenant toss PC.FarePolicy
+  case res of
+    Nothing -> return Nothing
+    Just fP -> fromCacType (fP, context, tenant, id, toss)
+
+instance FromCacType (BeamFP.FarePolicy, [(CacContext, Value)], String, Id FarePolicy, Int) Domain.FarePolicy where
```

Comment:
this instance duplicated a lot with fromTType instance
The only difference is sub queries

```
mFPPD <- CQueriesFPPD.getFPProgressiveDetailsFromCAC context tenant id' toss -- CAC
mFPPD <- QueriesFPPD.findById' (Id id) -- Beam
```

as further refactor we can create helper functions which will apply to both Beam and Cac instances
something like this

```
fromTTypeFarePolicyHelper :: Handler m -> BeamFP.FarePolicy -> m (Maybe FarePolicy)

instance FromCacType (BeamFP.FarePolicy, [(CacContext, Value)], String, Id FarePolicy, Int) Domain.FarePolicy where
  fromCacType (farePolicyT, context, tenant, id', toss) = fromTTypeFarePolicyHelper (cacFarePolicyHandler context tenant id' toss) farePolicyT

instance FromTType' BeamFP.FarePolicy Domain.FarePolicy where
  fromTType' farePolicyT = fromTTypeFarePolicyHelper beamFarePolicyHandler farePolicyT
```